clientapi/routing/joinroom.go

```diff
@@ -365,17 +375,63 @@ func (r joinRoomReq) joinRoomUsingServer(roomID string, server gomatrixserverlib
 		return nil, fmt.Errorf("r.federation.SendJoin: %w", err)
 	}
 
+	// A list of events that we have retried, if they were not included in
+	// the auth events supplied in the send_join.
+	retries := map[string]bool{}
+
+retryCheck:
 	if err = respSendJoin.Check(r.req.Context(), r.keyRing, event); err != nil {
-		return nil, fmt.Errorf("respSendJoin: %w", err)
+		switch e := err.(type) {
+		case gomatrixserverlib.MissingAuthEventError:
+			// Check that we haven't already retried for this event, prevents
+			// us from ending up in endless loops
+			if _, ok := retries[e.AuthEventID]; !ok {
+				// Ask the server that we're talking to right now for the event
+				tx, txerr := r.federation.GetEvent(r.req.Context(), server, e.AuthEventID)
+				if txerr != nil {
+					return nil, fmt.Errorf("r.federation.GetEvent: %w", txerr)
+				}
+				// For each event returned, add it to the auth events.
+				for _, pdu := range tx.PDUs {
+					ev, everr := gomatrixserverlib.NewEventFromUntrustedJSON(pdu, respMakeJoin.RoomVersion)
+					if everr != nil {
+						return nil, fmt.Errorf("gomatrixserverlib.NewEventFromUntrustedJSON: %w", everr)
+					}
+					respSendJoin.AuthEvents = append(respSendJoin.AuthEvents, ev)
+				}
+				// Mark the event as retried and then give the check another go.
+				retries[e.AuthEventID] = true
+				goto retryCheck
+			}
+			return nil, fmt.Errorf("respSendJoin (after retries): %w", e)
+		default:
+			return nil, fmt.Errorf("respSendJoin: %w", err)
+		}
 	}
 
-	if err = r.producer.SendEventWithState(
-		r.req.Context(),
-		gomatrixserverlib.RespState(respSendJoin.RespState),
-		event.Headered(respMakeJoin.RoomVersion),
-	); err != nil {
-		return nil, fmt.Errorf("r.producer.SendEventWithState: %w", err)
-	}
+	util.GetLogger(r.req.Context()).WithFields(logrus.Fields{
+		"room_id":          roomID,
+		"num_auth_events":  len(respSendJoin.AuthEvents),
+		"num_state_events": len(respSendJoin.StateEvents),
+	}).Info("Room join signature and auth verification passed")
+
+	// By this point we've verified all of the signatures, retrieved all of the
+	// missing auth events and verified that everything checks out. Nothing
+	// *should* go wrong in the roomserver after this point, so rather than have
+	// the client block on the roomserver taking in all of the new events, we
```

Comment:
I'm surprised that `r.producer.SendEventWithState` blocks. In polylith mode, that would just be writing to Kafka, and not waiting for any consumers. How does naffka handle this? If it is doing it synchronously, then we should fix *that* so we are consistent between mono/polylith modes.

Reply:
It definitely blocks in Naffka mode, I haven't tested it in polylith. I think we wait for errors.